streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java

```diff
@@ -108,7 +108,7 @@ public void testTimeTracking() {
         assertEquals(5, group.numBuffered());
         assertEquals(2, group.numBuffered(partition1));
         assertEquals(3, group.numBuffered(partition2));
-        assertEquals(2L, group.timestamp());
+        assertEquals(1L, group.timestamp());
```

Comment:
Sure, I will update the comments in PartitionGroup.

About whether we should still update stream time when filling in records to an empty partition, my gut feeling is yes. Consider the following example of two partitions in a task:

P1 (0): [0,1,2,3,4]
P2 (100): [100,101,102,103,104]

Again the digits above are record timestamps. When processing this task we will pop out records from P1 until it is empty. When a partition becomes empty, its stream time stays as the timestmap of the last popped record, hence P1's st (stream time) stays as 4, and hence the task st stays as 4.

P1 (4): []
P2 (100): [100,101,102,103,104]

Now consider we polled more records into P:

P1 (5): [5,6,7,8]
P2 (100): [100,101,102,103,104]

P1's st will be updated immediately as 5. If we do not update immediately, the task st is still 4, and only when record with timestamp 5 has been popped and processed, the task st will be "jumpping" from 4 to 6.

Of course, one can argue to favor "let the st jump in this case" over "let the st increase after adding new data, and hence users may observe st being increased without processing even a record yet". Personally I'm in favor of the latter, but I'm open to be convinced :)

Reply:
Another note: the above example is that we do not yet enforce processing before new records are enqueued to P1, otherwise it will look like

P1 (4): []
P2 (104): []

I.e. users processing records with ts 100+ while the stream time stays at 4. This is intentional in my design (users would be aware that some enforced processing is happening as st did not increment) but I'm also open for questions / arguments against it.

Reply:
It seems like it would be sensible either way.

I think the distinction really comes down to situations in which you observe the stream time *without* processing a record. I think that we only look at the stream time while we are processing a record, so there's actually no need to update the stream time during enqueue.

Actually, I didn't completely follow the "jumping" behavior in your example.

Starting from 
P1 (4): []
P2 (100): [100,101,102,103,104]
ST:4

we poll more records into P1:
P1 (5): [5,6,7,8]
P2 (100): [100,101,102,103,104]
ST:4

Then when we grab a record (5) to process, ST would advance to 5, no?

P1 (6): [6,7,8]
P2 (100): [100,101,102,103,104]
ST:5

In other words, a processor would observe:
record time = 0; stream time = 0
record time = 1; stream time = 1
record time = 2; stream time = 2
record time = 3; stream time = 3
record time = 4; stream time = 4
(processing pauses for a while while we wait for more records on P1)
record time = 5; stream time = 5
... 6, 7, ...
record time = 8; stream time = 8
(processing pauses for a while, until we give up on P1)
record time = 100; stream time = 8
record time = 101; stream time = 8
... etc (stream time doesn't advance any more until we see another record on P1)

I believe this sequence of observations is the same whether you advance stream time during enqueue or not.

Reply:
Got it. Thanks for the clarification!